docs/decisions/0015-outbox-pattern-and-production-modes.rst

```diff
@@ -0,0 +1,95 @@
+- Duplication becomes possible, so we would need a way to avoid sending the same event over and over again to the broker if the broker is failing to send acknowledgements. We may need to revisit existing events and improve documentation around ensuring that consumers can tolerate duplication, either by ensuring that events are idempotent or by keeping track of which event IDs have already been processed.
+- The database will be required to store an unbounded number of events during a broker outage, worker outage, or event bus misconfiguration.
+
+Rejected and Unplanned Alternatives
+***********************************
+
+Change Data Capture
+===================
+
+Change data capture (CDC) is a method of directly streaming database changes from one place to another by following the DB's transaction log. This provides the same transactionality benefits as the outbox method. `Debezium <https://debezium.io/>`_ is an example of such a system and can read directly from the database and produce to Kafka, where the data can then be transformed and routed to other systems. While a CDC platform could send data to the Open edX event bus, it would also be redundant with the event bus. In the example of Debezium, a deployment would still need a Kafka cluster even if they wanted to put event data into Redis.
+
+CDC systems also source their data at a lower level than we're targeting with the event bus; Django usually insulates us from schema details via an ORM layer, but CDC involves reading table data directly. We'd have tight coupling with our DB schemas. And the eventing system we've chosen to build operates at a higher abstraction layer than database writes, creating another conceptual mismatch. Theoretically, a CDC system could also be responsible for reading events from an outbox, allowing high-level eventing, but this is unlikely to be more palatable than just running a management command in a loop.
+
+Non-worker event production
+===========================
+
+The outbox pattern usually involves running a worker process that handles moving data from the outbox to the broker. However, it may be possible for deployers to avoid this with the use of some alternative middleware. For example, a custom middleware could flush events to the broker at the end of each event-producing request. The middleware's ``post_response`` would run outside of the request's main transaction. It would check if the request had created events, and if so, it would pull *at least that many* events from the outbox and produce them to the broker, then remove them from the outbox. If the server crashed before this could complete, later requests would eventually complete the work. This would also cover events produced by workers and other non-request-based processes.
```

Comment:
1. It seems like celery beat could be used as well. Same as it possibly could have been for event consumers.
2. Is an alternative also to just another type (like immediate), or allow outbox to work like immediate if some part of it is not configured? Logging would just need to be used for produce errors, as we would need to do now.

Reply:
1. I would group celery-beat under the umbrella of worker processes. (Most of our IDAs don't have a celery worker, right?)
2. I don't think there's any way for the adding-to-outbox side to know whether the removing-from-outbox side is properly configured.